File: ios/SlowThemDown/Views/Components/SpeedBadge.swift
import SwiftUI

struct SpeedBadge: View {
    let speed: Double
    let speedLimit: Double
    let system: MeasurementSystem

    private var category: SpeedCategory {
        guard speedLimit > 0 else { return .underLimit }
        let ratio = speed / speedLimit
        if ratio <= 1.0 { return .underLimit }
        if ratio <= 1.2 { return .marginal }
        return .overLimit
    }

    private var color: Color {
        switch category {
        case .underLimit: .green
        case .marginal: .yellow
        case .overLimit: .red
        }
    }

    var body: some View {
        Text("\(UnitConverter.displaySpeed(speed, system: system), specifier: "%.0f")")
            .font(.system(.title3, design: .rounded, weight: .bold))
            .foregroundStyle(.white)
            .padding(.horizontal, 12)
            .padding(.vertical, 6)
            .background(color, in: RoundedRectangle(cornerRadius: 8))
            .accessibilityElement(children: .combine)
            .accessibilityLabel("\(Int(UnitConverter.displaySpeed(speed, system: system))) \(UnitConverter.speedUnit(system)), \(category.label)")
    }
}

File: android/app/src/main/java/com/slowthemdown/android/ui/components/SpeedBadge.kt
package com.slowthemdown.android.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import com.slowthemdown.shared.model.MeasurementSystem
import com.slowthemdown.shared.model.SpeedCategory
import com.slowthemdown.shared.model.UnitConverter

@Composable
fun SpeedBadge(
    speed: Double,
    speedLimit: Double,
    system: MeasurementSystem,
    modifier: Modifier = Modifier,
) {
    val category = SpeedCategory.fromSpeed(speed, speedLimit)
    val color = when (category) {
        SpeedCategory.UNDER_LIMIT -> MaterialTheme.colorScheme.secondary
        SpeedCategory.MARGINAL -> MaterialTheme.colorScheme.tertiary
        SpeedCategory.OVER_LIMIT -> MaterialTheme.colorScheme.error
    }

    val displaySpeed = UnitConverter.displaySpeed(speed, system)
    val unit = UnitConverter.speedUnit(system)

    Box(
        modifier = modifier
            .clip(RoundedCornerShape(8.dp))
            .background(color.copy(alpha = 0.2f))
            .padding(horizontal = 12.dp, vertical = 6.dp)
    ) {
        Text(
            text = "%.1f %s".format(displaySpeed, unit),
            color = color,
            style = MaterialTheme.typography.labelLarge,
        )
    }
}
